src/jit/lowerxarch.cpp

```diff
@@ -2542,47 +2561,74 @@ bool Lowering::IsContainableHWIntrinsicOp(GenTreeHWIntrinsic* containingNode, Ge
-                            {
-                                supportsAlignedSIMDLoads   = !comp->canUseVexEncoding();
-                                supportsUnalignedSIMDLoads = !supportsAlignedSIMDLoads;
-                                supportsGeneralLoads       = supportsUnalignedSIMDLoads;
-
-                                break;
-                            }
+                            supportsSIMDScalarLoads = (ival <= 0x3F);
+                            supportsGeneralLoads    = supportsSIMDScalarLoads;
                         }
-
-                        assert(supportsAlignedSIMDLoads == false);
-                        assert(supportsUnalignedSIMDLoads == false);
-                        assert(supportsGeneralLoads == false);
+                        break;
                     }
-                    else
-                    {
-                        assert(supportsAlignedSIMDLoads == false);
-                        assert(supportsUnalignedSIMDLoads == false);
 
-                        supportsSIMDScalarLoads = true;
-                        supportsGeneralLoads    = supportsSIMDScalarLoads;
-                    }
+                    // We should only get here for integral nodes.
+                    assert(varTypeIsIntegral(node->TypeGet()));
+
+                    // These intrinsics require the contained operand to be the same
+                    // size as the baseType for the containing node.
+                    //
+                    // If it isn't, we can't mark the node as contained because it
+                    // might avoid a zero or sign-extension that should have otherwise
+                    // occurred.
+                    //
+                    // We need this even though there are overloads that explicitly
+                    // take byte/sbyte/int16/uint16, since the user may be calling
+                    // the int32/uint32 overload with an explicit cast.
+                    //
+                    // Finally, SIMDScalarLoads aren't supported because these intrinsics
+                    // take an actual scalar value, rather than taking a vector and only
+                    // operating on the lower bits.ther than taking a vector and only
+                    // operating on the lower bits.
+                    
+                    assert(supportsAlignedSIMDLoads == false);
+                    assert(supportsUnalignedSIMDLoads == false);
+                    assert(supportsSIMDScalarLoads == false);
+
+                    const unsigned expectedSize = genTypeSize(containingNode->gtSIMDBaseType);
+                    const unsigned operandSize  = genTypeSize(node->TypeGet());
 
+                    supportsGeneralLoads = (operandSize == expectedSize);
+
+                    if (!supportsGeneralLoads && ((expectedSize == 1) || (expectedSize == 2)))
+                    {
+                        // The scalar intrinsics that take an 8-bit or 16-bit operand
```

Comment:
Is there a better way to handle this?

Basically, this is checking for the case where we have a `byte`/`sbyte` or `short`/`ushort` that was already zero/sign-extended and was then stack-spilled as a 32-bit value....

This seems to be fairly common, especially when dealing with the `Vector128/256.Create(byte, ..., byte)` case (which takes 16 and 32 byte parameters)...

It seems, at least from the trees I saw, these are always `GT_LCL_VAR`. For example:
```
N007 (  6,  6) [000706] ------------              /--*  HWIntrinsic simd16 ubyte Insert $2cd
N004 (  1,  1) [000703] ------------              |  |     /--*  LIST      void   $205
N003 (  1,  1) [000702] ------------              |  |     |  \--*  CNS_INT   int    12 $45
N005 (  3,  3) [000704] ------------              |  |  /--*  LIST      void   <l:$a12, c:$a11>
N002 (  1,  1) [000701] ------------              |  |  |  \--*  LCL_VAR   int    V29 tmp19        u:2 (last use) <l:$7ad, c:$7ac>
N006 (  5,  5) [000705] ------------              |  \--*  LIST      void   <l:$a14, c:$a13>
N001 (  1,  1) [000700] ------------              |     \--*  LCL_VAR   simd16 V52 tmp42        u:13 (last use) $9e3
N009 (  6,  6) [000709] -A------R---              *  ASG       simd16 (copy) $VN.Void
N008 (  1,  1) [000707] D------N----              \--*  LCL_VAR   simd16 V52 tmp42        d:14
```

Reply:
I had looked in `ContainCheckBinary`, but it doesn't look like we do anything like this there, so I expect we either aren't doing this or it is handled elsewhere.

Reply:
Yes. For short fields, they show up as `GT_FIELD`, otherwise I only saw the `GT_CAST` (which is not containable).

Reply:
I'm a little confused:
- you should never see a `GT_FIELD` here; those are always transformed in morph.
- If you have a `GT_LCL_VAR` it is safe to load the full 4 bytes. It may be that these nodes are always retyped as `TYP_INT`, though I couldn't say that for certain. However, it's not entirely clear to me that this test is valid for the general case of `GT_IND`.

Can you give an example of a case you'd like to recognize where there's a size mismatch and it's not a `GT_LCL_VAR`?

Reply:
@CarolEidt, ah, I see. I was looking at the dump from before morph. It, in this situation, gets morphed from a `FIELD` (of an `ADDR->LCL_VAR`) to a `LCL_FLD`: https://gist.github.com/tannergooding/d566792973f9e8aaae1741089b4eee87#file-jitdump-txt-L415